extension/src/lib.rs

```diff
@@ -0,0 +1,304 @@
+            self.nulls.push(true);
+            return
+        }
+
+        let mut counter_correction = 0.0;
+        if self.is_counter {
+            let mut last_value = 0.0;
+            for (_, sample) in &self.window {
+                if *sample < last_value {
+                    counter_correction += last_value
+                }
+                last_value = *sample
+            }
+        }
+
+        let (latest_time, latest_val)= self.window.back().cloned().unwrap();
+        let (earliest_time, earliest_val) = self.window.front().cloned().unwrap();
+        let mut result_val = latest_val - earliest_val + counter_correction;
+
+        // all calculated durations and interval are in seconds
+        let mut duration_to_start = (earliest_time - self.current_window_min) as f64
+            / USECS_PER_SEC as f64;
+        let duration_to_end = (self.current_window_max - latest_time)  as f64
+            / USECS_PER_SEC as f64;
+
+        let sampled_interval = (latest_time - earliest_time) as f64
+            / USECS_PER_SEC as f64;
+        let avg_duration_between_samples = sampled_interval as f64
+            / (self.window.len()-1) as f64;
+
+        if self.is_counter && result_val > 0.0 && earliest_val >= 0.0 {
+            // Counters cannot be negative. If we have any slope at
+            // all (i.e. result_val went up), we can extrapolate
+            // the zero point of the counter. If the duration to the
+            // zero point is shorter than the durationToStart, we
+            // take the zero point as the start of the series,
+            // thereby avoiding extrapolation to negative counter
+            // values.
+            let duration_to_zero = sampled_interval * (earliest_val / result_val);
+            if duration_to_zero < duration_to_start {
+                duration_to_start = duration_to_zero
+            }
+        }
+
+        // If the first/last samples are close to the boundaries of the range,
+        // extrapolate the result. This is as we expect that another sample
+        // will exist given the spacing between samples we've seen thus far,
+        // with an allowance for noise.
+
+        let extrapolation_threshold = avg_duration_between_samples * 1.1;
+        let mut extrapolate_to_interval = sampled_interval;
+
+        if duration_to_start < extrapolation_threshold {
+            extrapolate_to_interval += duration_to_start
+        } else {
+            extrapolate_to_interval += avg_duration_between_samples / 2.0
+        }
+
+        if duration_to_end < extrapolation_threshold {
+            extrapolate_to_interval += duration_to_end
+        } else {
+            extrapolate_to_interval += avg_duration_between_samples / 2.0
+        }
+
+        result_val = result_val * (extrapolate_to_interval / sampled_interval);
+
+        if self.is_rate {
+            result_val = result_val / self.range as f64
+        }
```

Comment:
I'm used to writing code like this but the compiler suggests to use an if expression. I _assume_ this is to enable possible compiler optimizations.

Reply:
@antekresic nah, they're [completely equivalent](https://godbolt.org/z/-h2744), it's just a stylistic thing. Clippy (which is what I assume is firing here) has a bunch of style lints, some of which aren't even normative. Since this is a direct port of the go code, I think it's best to maintain similarly with the original over style